cty/msgpack/roundtrip_test.go

```diff
@@ -350,3 +350,58 @@ func TestRoundTrip(t *testing.T) {
 		})
 	}
 }
+
+// Unknown values with very long string prefix refinements do not round-trip
+// losslessly. If the prefix is longer than 256 bytes it will be truncated to
+// a maximum of 256 bytes.
+func TestRoundTrip_truncatesStringPrefixRefinement(t *testing.T) {
+	tests := []struct {
+		Value          cty.Value
+		Type           cty.Type
+		RoundTripValue cty.Value
+	}{
+		{
+			cty.UnknownVal(cty.String).Refine().StringPrefix(strings.Repeat("a", 1024)).NewValue(),
+			cty.String,
+			cty.UnknownVal(cty.String).Refine().StringPrefix(strings.Repeat("a", 255)).NewValue(),
+		},
+		{
+			cty.UnknownVal(cty.String).Refine().NotNull().StringPrefix(strings.Repeat("b", 1024)).NewValue(),
+			cty.String,
+			cty.UnknownVal(cty.String).Refine().NotNull().StringPrefix(strings.Repeat("b", 255)).NewValue(),
+		},
+		{
+			cty.UnknownVal(cty.String).Refine().StringPrefix(strings.Repeat("c", 255) + "-").NewValue(),
+			cty.String,
+			cty.UnknownVal(cty.String).Refine().StringPrefix(strings.Repeat("c", 255) + "-").NewValue(),
+		},
+		{
+			cty.UnknownVal(cty.String).Refine().StringPrefix(strings.Repeat("d", 255) + "🤷🤷").NewValue(),
```

Comment:
Happy to add comments for any of these that need it. For example, in this case the `StringPrefix()` call trims the final "🤷", then the remaining "🤷" is trimmed during serialisation.